zap/src/test/java/org/zaproxy/zap/authentication/FormBasedAuthenticationMethodTypeUnitTest.java

```diff
@@ -0,0 +1,248 @@
+        method.setAuthCheckingStrategy(AuthCheckingStrategy.POLL_URL);
+        method.setPollFrequencyUnits(AuthPollFrequencyUnits.REQUESTS);
+        method.setPollFrequency(5);
+        method.setLoggedInIndicatorPattern(LOGGED_IN_INDICATOR);
+
+        type = new FormBasedAuthenticationMethodType();
+        Mockito.when(method.getType()).thenReturn(type);
+
+        // Make sure no actual message processing is done
+        mockedSessionManagementMethod = Mockito.mock(SessionManagementMethod.class);
+        doNothing()
+                .when(mockedSessionManagementMethod)
+                .processMessageToMatchSession((HttpMessage) any(), (WebSession) any());
+
+        mockedContext = Mockito.mock(Context.class);
+        when(mockedContext.getSessionManagementMethod()).thenReturn(mockedSessionManagementMethod);
+
+        this.startServer();
+    }
+
+    @AfterEach
+    public void shutDownServer() throws Exception {
+        stopServer();
+    }
+
+    @Test
+    public void shouldReplaceUsernameInPollRequest() throws NullPointerException, IOException {
+        // Given
+        String test = "/shouldReplaceUsernameInPollRequest/test";
+        String encodedPattern =
+                URLEncoder.encode(
+                        PostBasedAuthenticationMethod.MSG_USER_PATTERN,
+                        StandardCharsets.UTF_8.name());
+        String pollUrl = "/shouldReplaceUsernameInPollRequest/pollUrl";
+        String pollData = "user=" + PostBasedAuthenticationMethod.MSG_USER_PATTERN;
+        String username = "user";
+        final List<String> orderedReqUrls = new ArrayList<>();
+        final List<String> orderedReqData = new ArrayList<>();
+
+        this.nano.addHandler(
+                new NanoServerHandler(pollUrl.replace(encodedPattern, username)) {
+                    @Override
+                    protected Response serve(IHTTPSession session) {
+                        orderedReqUrls.add(
+                                session.getUri() + "?" + session.getQueryParameterString());
+
+                        HashMap<String, String> map = new HashMap<String, String>();
+                        try {
+                            session.parseBody(map);
+                            orderedReqData.add(map.get("postData"));
+                        } catch (Exception e) {
+                        }
+                        return newFixedLengthResponse(LOGGED_IN_BODY);
+                    }
+                });
+        HttpMessage testMsg = this.getHttpMessage(test);
+        HttpMessage pollMsg = this.getHttpMessage(pollUrl + "?" + encodedPattern);
+
+        method.setPollUrl(pollMsg.getRequestHeader().getURI().toString());
+        method.setPollData(pollData);
+
+        User user = spy(new User(0, "user"));
+        user.setAuthenticationCredentials(
+                new UsernamePasswordAuthenticationCredentials(username, ""));
+        doReturn(mockedContext).when(user).getContext();
```

Comment:
It could just mock it (instead of instantiate and spy), e.g.:
```Java
    User user = mock(User.class);
    given(user.getAuthenticationCredentials()).willReturn(new UsernamePasswordAuthenticationCredentials(...));
    given(user.getContext()).willReturn(mockedContext);
```